---
title: "filtering_transcripts_by_expr_&_iso"
author: "Holly"
date: "September 28, 2020"
output: html_document
---
This script provides a workflow to filter out low expression transcriptomics using Trinity's filter_low_expr_transcripts.pl to produce a streamlined TMM.matrix to use in WGCNA analysis (or wherever else it may be used).

##1. First, we will work from the command line to filter our transcripts based on highest(longest) isoform. Save the output from this function in a new file (it will be in fasta format and can be saved as a .fasta or .txt). We used our TMM.matrix file produced from RSEM as our matrix input for this command, as it provides normalized expression values for our list of transcripts.
```{r}
perl /projects/rockfish/transcriptomics/bake1349/trinityrnaseq-Trinity-v2.4.0/util/filter_low_expr_transcripts.pl --matrix ../All_DGE/matrix.TMM.EXPR.matrix --transcripts /projects/rockfish/transcriptomics/EXP11_ScarnSmel_fluct_gill/Copgil_mapping/Copgil_HMD_Trinity_FR.fasta --highest_iso_only --trinity_mode > Black_normalized_filtered.txt

#note: you can also filter isoforms in this program with the argument --min_pct_dom_iso <int> and it will filter isoforms based on  a minimum person of dominant isoform expression that you set (ex: 10%). This command is mutually exclusive with the other filter, so you use one or the other. 

```

##2. Next, using our new file containing filtered isoforms as transcript input, we will filter again based on expression. However, the transcripts in this file still retain their _iX isoform endings, and the transcripts in our TMM.matrix file do not. Unfortunately this version of the program does not seem to be able to recognize the transcript without the isoform ending, so we must first remove them. (This should be okay as our TMM file was already produced using the longest isoforms, and thus the transcripts in each list should be unique and matching). To filter, we use: 

```{r}
sed 's/_i../ /' Black_normalized_filtered.txt > Black_normalized_filtered_no_ix.txt
```


Then, check that you retained the correct number of transcripts by checking the 'TRINITY' count: 
```{r}
awk '/TRINITY*/ ' Black_normalized_filtered.txt | wc
          -> 420348
awk '/TRINITY*/ ' Black_normalized_filtered_no_ix.txt | wc -l
          -> 420348
```


Now we can filter expression using our new no endings file:
```{r}
perl /projects/rockfish/transcriptomics/bake1349/trinityrnaseq-Trinity-v2.4.0/util/filter_low_expr_transcripts.pl --matrix ../All_DGE/matrix.TMM.EXPR.matrix --transcripts Black_normalized_filtered_no_ix.txt --min_expr_any 1 --trinity_mode > Black_normalized_filtered_expr.txt
    #You can play around with the expression threshold, i.e. 0.5 vs 1
    # ->  Retained 98432 / 420348 = 23.42% of total transcripts.
```
At this point you will want to pull your filtered file to your desktop (and your TMM file if ou haven't already) and convert the .txt files to csv files in excel. 


##3. Now that we have our fully filtered transcript list, we can use R to merge the filtered list with our original TMM.matrix to obtain corresponding expression values to use for our WGCNA Analysis
